src/main.cpp

```diff
@@ -2799,20 +2799,40 @@ static int64_t nTimeFlush = 0;
 static int64_t nTimeChainState = 0;
 static int64_t nTimePostConnect = 0;
 
+struct CompareBlockIndexByWork {
+    bool operator()(const CBlockIndex* a, const CBlockIndex* b) const {
+        return a->nChainWork < b->nChainWork;
+    }
+};
+
+// Used in ActivateBestChain/ConnectBlock to keep track of blocks connected
+// and notify listeners in the right order.
+// Note that since mapBlocks is sorted lowest-work to highest-work, iterating
+// over it will give you the order in which the blocks were connected
+struct BlockConnectTrace {
+    std::map<CBlockIndex*, std::shared_ptr<const CBlock>, CompareBlockIndexByWork> mapBlocks;
```

Comment:
I think it may be worthwhile for you to make this a class and keep the implementation details hidden...